onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -185,20 +241,35 @@ common::Status SaveInitializedTensors(
   }
 
   // tensors requiring a specific allocation order are traced first, to ensure they are allocated in order
+  // NB1: vector with init allocation order may contain a subset of all tensors (or none at all)
+  // NB2: only skip tracing and planing memory when data is external (i.e mmap) and on CPU.
+  //    when data is external and on GPU, need to copy first to cpu memory, then to gpu memory.
   auto initialized_tensors_to_allocate = id_to_initialized_tensor;
   for (int ort_value_index : initializer_allocation_order) {
     const auto entry = initialized_tensors_to_allocate.find(ort_value_index);
-    // can not trace string tensor
-    ORT_ENFORCE(entry != initialized_tensors_to_allocate.end() && entry->second->data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING);
-    ORT_RETURN_IF_ERROR(planner.Trace(entry->first, entry->second));
+    auto location = exec_plan.GetLocation(ort_value_index).name;
+    if (utils::HasExternalData(*entry->second) && (strcmp(location, CPU) == 0)) {
+      // external data will be memory mapped, no need to plan for its allocation
+      continue;
+    } else {
+      // can not trace string tensor
+      ORT_ENFORCE(entry != initialized_tensors_to_allocate.end() &&
+                  entry->second->data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING);
+      ORT_RETURN_IF_ERROR(planner.Trace(entry->first, entry->second));
+    }
     initialized_tensors_to_allocate.erase(entry);
   }
 
   for (const auto& entry : initialized_tensors_to_allocate) {
+    auto location = exec_plan.GetLocation(entry.first).name;
     // We don't want to trace shared initializers since their memory is provided by the user
     if (user_supplied_initializer_ids.find(entry.first) != user_supplied_initializer_ids.end()) {
       continue;
     }
+    if (utils::HasExternalData(*entry.second) && (strcmp(location, CPU) == 0)) {
+      // exernal data will be memory mapped, no need to plan for its allocation
```

Comment:
nit: exernal -> external.

Is the data going to be memory mapped, or is it already memory mapped?

Reply:
fixed the comment